### PR TITLE
Add kubernetes configuration source

### LINKED ISSUE #169: Multi-tenant mode config
- Watch directory for tenant configurations, lazy load config on request
    - `/index.json`: a JSON object with host as key, and app ID as value
    - `/{app ID}/`: directory containing the configuration for the app
- Single tenant mode: configure a directory for app config, instead of individual config files
    - Templates are resolved relative to the `authgear.yaml` file.

### PR DESCRIPTION
**Pull request by kiootic**
fix #169

### CODE REVIEW

File: pkg/lib/config/configsource/kubernetes.go [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+	"k8s.io/client-go/util/homedir"
+
+	"github.com/authgear/authgear-server/pkg/lib/config"
+	"github.com/authgear/authgear-server/pkg/util/log"
+)
+
+type KubernetesLogger struct{ *log.Logger }
+
+func NewKubernetesLogger(lf *log.Factory) KubernetesLogger {
+	return KubernetesLogger{lf.New("kubernetes-config")}
```
Comment (louischan-oursky) [inline pkg/lib/config/configsource/kubernetes.go:38]: The logger name should be `configsource-kubernetes`

File: pkg/lib/config/configsource/kubernetes.go [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+}
+
+type Kubernetes struct {
+	Logger LocalFSLogger
```
Comment (louischan-oursky) [inline pkg/lib/config/configsource/kubernetes.go]: Should be `KubernetesLogger`

File: pkg/lib/config/configsource/kubernetes.go [hunk 3]
```diff
@@ -91,6 +91,7 @@ func (k *Kubernetes) Open() error {
 	secretCtrl := k.newController(corev1.ResourceSecrets, &corev1.Secret{}, k.onUpdate, k.onDelete)
 	go configMapCtrl.Run(done)
 	go secretCtrl.Run(done)
+	go k.cleanupCache(done)
```
Comment (louischan-oursky) [inline pkg/lib/config/configsource/kubernetes.go:94]: Consider using [this library](https://github.com/allegro/bigcache) instead? It has max number of entries, max cache size in bytes support.

File: pkg/lib/config/configsource/config.go [hunk 4]
```diff
@@ -22,8 +22,8 @@ type Config struct {
 	KubeNamespace string `envconfig:"KUBE_NAMESPACE"`
 	// KubeAppHostMapName indicates the name of app host mapping ConfigMap
 	KubeAppHostMapName string `envconfig:"KUBE_APP_HOST_MAP_NAME" default:"app-hosts"`
-	// KubeAppConfigPattern sets the RegEx that matches the app ID in ConfigMap/Secret name
-	KubeAppConfigPattern string `envconfig:"KUBE_APP_CONFIG_PATTERN" default:"^app-data-(.+)$"`
+	// KubeAppConfigPrefix sets the name prefix of app ConfigMap/Secret name
+	KubeAppConfigPrefix string `envconfig:"KUBE_APP_CONFIG_PREFIX" default:"app-data-"`
```
Comment (louischan-oursky) [inline pkg/lib/config/configsource/config.go:26]: Should we use labels instead to identity our ConfigMap and Secret?

File: pkg/lib/config/configsource/kubernetes.go (hunk 3 above)
Comment (kiootic) [reply pkg/lib/config/configsource/kubernetes.go:94]: It require byte slices as value, which make the current behavior (fetch at most once, even if there are concurrent requests) impossible.

File: pkg/lib/config/configsource/config.go (hunk 4 above)
Comment (kiootic) [reply pkg/lib/config/configsource/config.go:26]: Sure, we can use labels. So:
- A label to indicate the host mapping ConfigMap, e.g. `https://authgear.com/kubernetes/host-mapping: true`
- A label to indicate the app config ConfigMap & Secret, e.g. `https://authgear.com/kubernetes/config-app-id: 9C390E3B-9030-48B0-A350-57F51572145F`

Both label names are constants and cannot be configured. Does that sounds good to you?

**Comment by louischan-oursky** [comment]
I update the config map to remove login ID key and then refresh the page. There is a log `invalidated cached config` but the login ID is still show removed login ID key.

**Comment by louischan-oursky** [comment]
It turned out my configuration was correct.

File: pkg/lib/config/configsource/kubernetes.go [hunk 5]
```diff
@@ -30,6 +29,11 @@ import (
 	"github.com/authgear/authgear-server/pkg/util/log"
 )
 
+const (
+	LabelHostMapping = "authgear.com/host-mapping"
```
[the remark this answers is not in the document]
Comment (kiootic) [reply pkg/lib/config/configsource/kubernetes.go:33]: Label keys can have at most one slash: https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set

[the remark this answers is not in the document]
Comment (louischan-oursky) [reply pkg/lib/config/configsource/kubernetes.go:33]: Didn't know that! 😳

[the remark this answers is not in the document]
Comment (louischan-oursky) [reply pkg/lib/config/configsource/kubernetes.go:33]: What about `kubernetes.authgear.com/host-mapping`?

[the remark this answers is not in the document]
Comment (kiootic) [reply pkg/lib/config/configsource/kubernetes.go:33]: Sure that's valid, but do we need to include 'kubernetes' somewhere in the label key? I think the context makes it pretty clear already.

[the remark this answers is not in the document]
Comment (louischan-oursky) [reply pkg/lib/config/configsource/kubernetes.go:33]: Fair enough.

**Pull request merged**